src/scale_impls/decode.rs

```diff
@@ -13,239 +13,314 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use super::TypeId;
+use core::marker::PhantomData;
+
 use crate::prelude::*;
 use crate::value_type::{Composite, Primitive, Value, ValueDef, Variant};
-use scale_decode::FieldIter;
-use scale_info::{form::PortableForm, Path, PortableRegistry};
+use scale_decode::{FieldIter, TypeResolver};
+use scale_info::PortableRegistry;
 
 // This is emitted if something goes wrong decoding into a Value.
 pub use scale_decode::visitor::DecodeError;
 
-/// Decode data according to the [`TypeId`] provided.
+/// Decode data according to the type id provided.
 /// The provided pointer to the data slice will be moved forwards as needed
 /// depending on what was decoded.
 pub fn decode_value_as_type(
     data: &mut &[u8],
-    ty_id: TypeId,
+    ty_id: u32,
     types: &PortableRegistry,
-) -> Result<Value<TypeId>, DecodeError> {
-    scale_decode::visitor::decode_with_visitor(data, ty_id, types, DecodeValueVisitor)
+) -> Result<Value<u32>, DecodeError> {
+    scale_decode::visitor::decode_with_visitor(
+        data,
+        &ty_id,
+        types,
+        // note: in this case the `FromMapper` converts the u32 into a u32, and is just an identity mapping.
+        DecodeValueVisitor::<PortableRegistry, u32, FromMapper>::new(),
+    )
+}
+
+/// A more generic version of [`crate::scale::decode_as_type`], but in most cases you should
+/// just use [`crate::scale::decode_as_type`] because then no type params need to be configured.
+pub fn decode_any_value_as_type<T, R: TypeResolver>(
+    data: &mut &[u8],
+    ty_id: R::TypeId,
+    types: &R,
+) -> Result<Value<T>, DecodeError>
+where
+    T: From<R::TypeId>,
```

Comment:
What's the value of this? Offhand I'd be inclined to just return `Value<R::TypeId>` (and if somebody wants to change the context value, there are fns like https://github.com/paritytech/scale-value/blob/1a404c1b5e023587ef11b5993afc97adcdc11174/src/value_type.rs#L127 available). But perhaps there's a concrete reason we need this that I'm missing :)

If we didn't have the `From` (which would require type info like `let foo: Value<Something> = decode_any_value_as_type(..)`) then I also would think that we don't need to expose this _and_ `decode_value_as_type`; `decode_value_as_type` can just be more generic and it shouldn't lead to any extra type info being needed etc.